### PR TITLE
Issue 20: Add example specification SLPv2

### PR DESCRIPTION
**Pull request by urbasus**
Example specification for SLPv2. Fixes issue #20.

Open TODOs:
- [x] ~~Use member 'Size instead of hard-coded offsets in size arithmetic~~
   - Rejected: Would improve neither readability nor maintainability
   - ~~Should use <Field>'Position or similar. No such thing exists in current version of RecordFlux~~
   - Using Field'First and Field'Last, as suggested by [this issue comment](https://github.com/Componolit/RecordFlux-specifications/issues/20#issuecomment-753372283)
- [x] ~~Constrain XID to =0 when Function_Id = DIRECTORY_AGENT_ADVERTISEMENT~~
   - Rejected: Unclear meaning of RFC statement "Only unsolicited DAAdverts are sent with an XID of 0"
- [x] Review RFC for missed constraints
   - All string format constraints are omitted
   - Size of some formatted strings are known to be of a minimum size, such as URL. Such constraints are omitted since they would be redundant with necessary, application provided validation
- [x] Fix Service_Location_Extension list OR drop from specification (Opaque)
   - Extensions are provided by a single opaque field
- [x] Test
- [x] Style fixes

### CODE REVIEW

**Comment by urbasus** [comment]
All done, no more changes planned.

**Comment by senier** [comment]
Thank you for this contribution. I like the style and consistency of the result!

We won't review and merge the PR just now, as there is another PR in the queue (#13 addressing #7) which I'd like to merge first. This said PR will introduce a tool to validate a specification against a set of known-good and known-bad message samples. Our goal is to have a decent set of validation samples for every specification in the repository. In the future we'd even like to measure the (specification) coverage to make sure samples for all branches of a spec are available (cf. #11). It would be great if you could provide some validation samples for SLP once #13 is merged.

Furthermore, we'll need a signed contributors agreement. I'll send you a separate email on this shortly.

**Comment by senier** [comment]
@urbasus, we are ready to merge your PR now. Thank you for your patience ;) Could you please rebase it to the current `main` branch? The main difference you will notice is the integration and automatic execution of our new validation tool.

It would be great if you could provide some samples for SLP. Those would have to be raw binaries which go into

`tests/data/<protocol_name_in_lower_case>/<message_name_in_lower_case>/{valid|invalid}`

As you would expect, the `valid` directory contains message that have to be parsed successfully and the `invalid` directory those that must be rejected. No worries if you don't have invalid messages at hand or only few valid ones. This is OK (we are currently building coverage analysis into the validation tool and want to enforce the coverage level in the future - currently it's no error to have no test cases though).

Don't hesitate to get in touch if you have any questions.

**Comment by urbasus** [comment]
Good to hear.

I'm having some technical issues at the moment. Might take a few more weeks before I can get it done.

**Comment by senier** [comment]
> Good to hear.
> 
> I'm having some technical issues at the moment. Might take a few more weeks before I can get it done.

No problem. Take your time - we're not in a hurry.

**Comment by urbasus** [comment]
I've added a few messages.

The validation does not pass:

```
tests/data/slpv2/service_request/valid/srvreq1.raw: FAILED
FalseNegative
pyrflx: error: Bitstring representing the message is too short - stopped while parsing field: PR_List_String

tests/data/slpv2/service_request/valid/srvreq2.raw: FAILED
FalseNegative
pyrflx: error: Bitstring representing the message is too short - stopped while parsing field: PR_List_String

tests/data/slpv2/service_request/valid/srvreq3.raw: FAILED
FalseNegative
pyrflx: error: Bitstring representing the message is too short - stopped while parsing field: PR_List_String
```

Looking at the test data the Length_Of_PR_List field determining the length of the PR_List_String field is supposed to be zero. The error message suggests it is not. Meaning something is wrong with the spec or the generator.

I cannot seem to find the problem. Would appreciate anyone taking a look at it.

Running rflx version 0.5.0-pre, installed by `pip3 install RecordFlux@git+https://github.com/Componolit/RecordFlux.git@fc40c52`

**Comment by senier** [comment]
> I cannot seem to find the problem. Would appreciate anyone taking a look at it.

The `Length_Of_PR_List` field of `Service_Request` is 16 bit wide, whereas the first two bytes of `srvreq1.raw` is `02 01`. This would mean that `PR_List_String` has a length of 513 bytes which clearly exceeds the size of you sample. The parser is correct here. Either the length field in the spec should be smaller or your data doesn't match the spec.

**Comment by urbasus** [comment]
Ok, I see. The data is header+payload but the testing expects payload only. Makes sense. Thanks.

Hämta Outlook för Android<https://aka.ms/ghei36>

________________________________

**Comment by urbasus** [comment]
On fixing the validation message paths I discovered errors in the spec due to having accidentally mixing bit and byte counts. After fixing those validation now passes.

On fixing those errors however the spec tests themselves do not pass because of field size being potentially negative:

```
-- 9.2. Authentication Blocks
type Authentication_Block_Authentication_Block_Length is range 10 .. 2**16-1 with Size => 16;
type Authentication_Block is
   message
      Block_Structure_Descriptor : U16;
      Authentication_Block_Length : Authentication_Block_Authentication_Block_Length;
      Authentication_Block_Timestamp : U32;
      SLP_SPI_String_Length : U16;
      SLP_SPI_String : Opaque with Size => SLP_SPI_String_Length * 8;
      -- error msg: lx:102:10: model: error: negative size for field "Structured_Authentication_Block" (Block_structure_Descriptor -> Authentication_Block_Length -> Authentication_Block_Timestamp -> SLP_SPI_String_Length -> SLP_SPI_String -> Structured_Authentication_Block)
      Structured_Authentication_Block : Opaque
         with Size => 8 * Authentication_Block_Length - (SLP_SPI_String'Last - Block_Structure_Descriptor'First + 1);
   end message;
type Authentication_Blocks is array of Authentication_Block;
```

Just a few more moles to whack...

**Comment by urbasus** [comment]
I think I got it. Ready for merge.

**Comment by urbasus** [comment]
Rebased on top of main. Ready for merge, again.

File: slpv2.rflx [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+   type URL_Entry is
+      message
+         Reserved : URL_Entry_Reserved;
+         Lifetime : U16;
+         URL_Length : U16;
+         URL : Opaque with Size => URL_Length * 8;
+         Number_Of_URL_Authentications : U8;
+         Authentication_Blocks : Authentication_Blocks
+            with Size => Number_Of_URL_Authentications * 8;
+      end message;
```
Comment (treiher) [inline slpv2.rflx:120]: Please use a consistent style for the field aspects in all message types. All instances should be formatted like `Authentication_Blocks` (`with` on a separate line).

File: slpv2.rflx [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+   type Authentication_Block is
+      message
+         Block_Structure_Descriptor : U16;
+         Authentication_Block_Length : Authentication_Block_Authentication_Block_Length;
+         Authentication_Block_Timestamp : U32;
+         SLP_SPI_String_Length : U16;
+         SLP_SPI_String : Opaque with Size => SLP_SPI_String_Length * 8;
+         Structured_Authentication_Block : Opaque
+            with Size => 8 * Authentication_Block_Length - (SLP_SPI_String'Last - Block_Structure_Descriptor'First + 1)
+               if (8 * Authentication_Block_Length >= (SLP_SPI_String'Last - Block_Structure_Descriptor'First + 1));
```
Comment (treiher) [inline slpv2.rflx:105]: The condition (`if`) should have the same indentation as the size aspect (`with`).

File: slpv2.rflx [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+         Attribute_List : Opaque with Size => Length_Of_Attribute_List_String * 8;
+         Number_Of_Authentication_Blocks : U8;
+         Authentication_Blocks : Authentication_Blocks
+            with Size => Number_Of_Authentication_Blocks * 8;
+      end message;
+   for SLPv2 use (Payload => Service_Agent_Advertisement)
+      if Function_ID = FI_SA_ADVERTISEMENT;
+
+   -- 9.1. Service Location Protocol Extensions
+   -- TODO: Support verification of extensions; not possible/feasible with RecordFlux 5.0, see issue comment https://github.com/Componolit/RecordFlux-specifications/issues/20#issuecomment-752084366
```
Comment (treiher) [inline slpv2.rflx:195]: I have created an issue to track that limitation. Please replace the comment by a reference to the issue:

```
-- ISSUE: Componolit/RecordFlux#598
```

**Review by treiher — CHANGES_REQUESTED** [review]
Sorry for the delay. Apart from some minor style issues the specification looks good to me. As soon as these are fixed, the PR can be merged.

File: slpv2.rflx (hunk 1 above)
Comment (urbasus) [reply slpv2.rflx]: Fixed.

File: slpv2.rflx (hunk 2 above)
Comment (urbasus) [reply slpv2.rflx]: Fixed.

File: slpv2.rflx (hunk 3 above)
Comment (urbasus) [reply slpv2.rflx]: Fixed.

**Review by treiher — APPROVED** [review]
LGTM. Thanks for your contribution!

**Pull request merged**